test/unit/html/HtmlConverterTest.php

```diff
@@ -108,6 +108,280 @@ public function testinsertSnippetAndHreflangTags()
         $this->assertEquals($expected_html, $translated_html);
     }
 
+    public function testinsertSnippetAndHreflangTagsWithCustomAlias()
+    {
+        $html = '<html><body><a>hello</a></body></html>';
+        $settings = array(
+            'supported_langs' => array('fr'),
+            'default_lang' => 'en',
+            'custom_lang_aliases' => array('en' => 'custom_en'),
+            'lang_param_name' => 'wovn',
+            'url_pattern_name' => 'path'
+        );
+        list($store, $headers) = StoreAndHeadersFactory::fromFixture('default', $settings);
+        $converter = new HtmlConverter($html, 'UTF-8', $store->settings['project_token'], $store, $headers);
+        list($translated_html) = $converter->insertSnippetAndHreflangTags(false);
+
+        $expected_html = "<html><body><link rel=\"alternate\" hreflang=\"fr\" href=\"http://my-site.com/fr/\"><script src=\"//j.wovn.io/1\" data-wovnio=\"key=123456&amp;backend=true&amp;currentLang=en&amp;defaultLang=en&amp;urlPattern=path&amp;langCodeAliases={&quot;en&quot;:&quot;custom_en&quot;}&amp;langParamName=wovn\" data-wovnio-info=\"version=WOVN.php_VERSION\" async></script><a>hello</a></body></html>";
+        $this->assertEquals($expected_html, $translated_html);
+    }
+
+    /**
+     * Call protected/private method of a class.
+     *
+     * @param object &$object    Instantiated object that we will run method on.
+     * @param string $methodName Method name to call
+     * @param array  $parameters Array of parameters to pass into method.
+     *
+     * @return mixed Method return.
+     */
+    public function invokeMethod(&$object, $methodName, array $parameters = array())
+    {
+        $reflection = new \ReflectionClass(get_class($object));
+        $method = $reflection->getMethod($methodName);
+        $method->setAccessible(true);
+
+        return $method->invokeArgs($object, $parameters);
+    }
+
+    public function testBuildHrefLangPath()
+    {
+        $html = '<html><body><a>hello</a></body></html>';
+        $settings = array(
+            'supported_langs' => array('ja'),
+            'default_lang' => 'en',
+            'lang_param_name' => 'wovn',
+            'url_pattern_name' => 'path'
+        );
+        list($store, $headers) = StoreAndHeadersFactory::fromFixture('default', $settings);
+        $converter = new HtmlConverter($html, 'UTF-8', $store->settings['project_token'], $store, $headers);
+        $expected_href = 'http://my-site.com/ja/';
+        $generated_href = $this->invokeMethod($converter, 'buildHrefLang', array('ja'));
```

Comment:
I don't really agree with testing private methods. Is this really the easiest way to test it right now?

Reply:
I'm doing this mostly to bring some level of confidence into the implementation. The chain of private method calls are too long before arriving at the public method making me feel it is not quite enough to just test the public ones. Also the private methods often have multiple behaviors, without testing them directly it's hard to figure out which behavior went wrong when the code breaks.

Reply:
That's more of a problem with the class itself being too big, but tests are better than no tests, certainly. 

For this case I would just be calling the convert method and checking that the resulting HTML contains the expected hreflangs, for example.

Reply:
Fair point. I'll move the individual tests on private methods to `insertSnippetAndHreflangTags`.